From c8515c09ab23b7f88171223ccb4c6793fd45d470 Mon Sep 17 00:00:00 2001
From: Myy <myy@miouyouyou.fr>
Date: Sun, 25 Jun 2017 15:38:05 +0000
Subject: [PATCH] The tinkerboard boot fix AGAIN.

This time the fix is protected by a new boot parameter. You'll have to
pass the following option to the kernel to enable this fix :

rockchip_tinkerrebootfix=on

Signed-off-by: Myy <myy@miouyouyou.fr>
---
 drivers/mmc/host/dw_mmc-rockchip.c | 32 ++++++++++++++++++++++++++++++++
 1 file changed, 32 insertions(+)

diff --git a/drivers/mmc/host/dw_mmc-rockchip.c b/drivers/mmc/host/dw_mmc-rockchip.c
index 372fb6e..7a89e99 100644
--- a/drivers/mmc/host/dw_mmc-rockchip.c
+++ b/drivers/mmc/host/dw_mmc-rockchip.c
@@ -12,9 +12,11 @@
 #include <linux/clk.h>
 #include <linux/mmc/host.h>
 #include <linux/of_address.h>
+#include <linux/regulator/consumer.h>
 #include <linux/mmc/slot-gpio.h>
 #include <linux/pm_runtime.h>
 #include <linux/slab.h>
+#include <linux/string.h>
 
 #include "dw_mmc.h"
 #include "dw_mmc-pltfm.h"
@@ -380,6 +382,36 @@ static struct platform_driver dw_mci_rockchip_pltfm_driver = {
 	},
 };
 
+/* A weird reboot hack
+ * Discussed here :
+ * https://github.com/Miouyouyou/MyyQi/issues/8
+ */
+static void dw_mci_rockchip_platfm_shutdown(struct platform_device *pdev)
+{
+	struct dw_mci *host = platform_get_drvdata(pdev);
+	struct mmc_host *mmc = host->cur_slot->mmc;
+
+	if (!IS_ERR(mmc->supply.vqmmc))
+		regulator_set_voltage(mmc->supply.vqmmc, 3000000, 3300000);
+}
+
+static int dw_mci_rockchip_setup_tinker_reboot_fix(char *str)
+{
+	if (strstr(str, "on")) {
+		printk(KERN_ERR "( Myy ) Enabling Tinkerboard's reboot fix !\n");
+		dw_mci_rockchip_pltfm_driver.shutdown = 
+			dw_mci_rockchip_platfm_shutdown;
+	}
+	else 
+		printk(
+			KERN_INFO "( Myy ) Not enabling Tinkerboard's reboot fix\n"
+		);
+	
+	return 0;
+}
+
+__setup("rockchip_tinkerrebootfix=", dw_mci_rockchip_setup_tinker_reboot_fix);
+
 module_platform_driver(dw_mci_rockchip_pltfm_driver);
 
 MODULE_AUTHOR("Addy Ke <addy.ke@rock-chips.com>");
-- 
2.10.2

